src/metabase/pulse/render/datetime.clj

```diff
@@ -11,33 +11,107 @@
            java.time.Period
            java.time.temporal.Temporal))
 
+(defn temporal-string?
+  "Returns `true` if the string `s` is parseable as a datetime.
+
+  `(temporal-string? \"asdf\")` -> false
+  `(temporal-string? \"2020-02-02\")` -> true"
+  [s]
+  (boolean
+    (try
+      (u.date/parse s)
+      (catch Exception _e false))))
+
 (defn- reformat-temporal-str [timezone-id s new-format-string]
   (t/format new-format-string (u.date/parse s timezone-id)))
 
+(defn- day-of-week
+  [s abbreviate]
+  (get-in
+    {"1" {:full (tru "Sunday") :abbreviated (tru "Sun")}
+     "2" {:full (tru "Monday") :abbreviated (tru "Mon")}
+     "3" {:full (tru "Tuesday") :abbreviated (tru "Tue")}
+     "4" {:full (tru "Wednesday") :abbreviated (tru "Wed")}
+     "5" {:full (tru "Thursday") :abbreviated (tru "Thu")}
+     "6" {:full (tru "Friday") :abbreviated (tru "Fri")}
+     "7" {:full (tru "Saturday") :abbreviated (tru "Sat")}}
+    (if abbreviate
+      [s :abbreviated]
+      [s :full])))
+
+(defn- month-of-year
+  [s abbreviate]
+  (get-in
+    {"1"  {:full (tru "January") :abbreviated (tru "Jan")}
+     "2"  {:full (tru "February") :abbreviated (tru "Feb")}
+     "3"  {:full (tru "March") :abbreviated (tru "Mar")}
+     "4"  {:full (tru "April") :abbreviated (tru "Apr")}
+     "5"  {:full (tru "May") :abbreviated (tru "May")}
+     "6"  {:full (tru "June") :abbreviated (tru "Jun")}
+     "7"  {:full (tru "July") :abbreviated (tru "Jul")}
+     "8"  {:full (tru "August") :abbreviated (tru "Aug")}
+     "9"  {:full (tru "September") :abbreviated (tru "Sep")}
+     "10" {:full (tru "October") :abbreviated (tru "Oct")}
+     "11" {:full (tru "November") :abbreviated (tru "Nov")}
+     "12" {:full (tru "December") :abbreviated (tru "Dec")}}
+    (if abbreviate
+      [s :abbreviated]
+      [s :full])))
+
+(defn- x-of-y
+  "Format an integer as x-th of y, for example, 2nd week of year."
+  [int-str]
+  (let [n (int (read-string int-str))]
+    (case (mod n 10)
+      1 (tru "{0}st" n)
+      2 (tru "{0}nd" n)
+      3 (tru "{0}rd" n)
+      (tru "{0}th" n))))
+
+(defn- hour-of-day
+  [s time-style]
+  (let [n (int (read-string s))
```

Comment:
`read-string` is 100% wrong. If you want to get a numeric value use a proper parse function. Clojure 1.11.x now has `parse-long`, or static methods from java `(Long/parseLong "123")`

Reply:
Ah, yes. Thanks for calling that out. Felt wrong at the time, then I forgot to try dig in more. I didn't know about `parse-long`, that's good to know!